### PR TITLE
Further ranged cleanup (part 2) [RDY]

### PR DESCRIPTION
**Pull request by mugling**
Continues from #15936.

`gun_base` now provides default values for `color` and `symbol`. The previous ad-hoc values were inconsistent but `grey (` was by far the most common. It is entirely possible (and probably desirable) to have a different symbol for each gun class but I don't play curses much so I'll let someone else propose the values.

As per @Coolthulhu's suggestion the improvised weapons become much less useless. They now derive their stats from the closest match with suitable penalties including increased `weight`, `volume` and `reload` along with markedly reduced `durability` and accept a much reduced set of modifications to maintain balance. They are serviceable but always strictly inferior to their base type.

The 12 gauge pistol is migrated and item drops added whilst the shotgun revolver becomes an unwieldy pistol (e4389d7) to differentiate it from the otherwise identical lever shotgun (84571f4)

All migrated guns are tagged with an class with four new classes (`rifle_pipe`, `rifle_machine`, `rifle_carbine` and `smg_compact`) added by cb2412c. Machine guns gain a bonus to recoil via their class in 0432380 which helps make them less useless (although ammo belts still need work). Pipe rifles will be further balanced once I address overall recoil, range and dispersion for ammo as a whole so their class is mainly provided as an extension point for now.

A further PR will aim to make `aim_speed` more relevant (and probably rename it to `aim_cost`). The Rivtech weapons, small holdout pistols and `compact_smg` are atypically good for their respective classes (81375d3).

Again when gun reliability is reworked `durability` (or lack of) will be more significant. This PR applies some sanity so that most guns use `durability 8` with exceptions for improvised weapons (typically `durability -2`) or where their `description` suggests otherwise (9b59163). The Rivtech weapons have above average `durability` whilst many Leadworks weapons (excluding L523 base) become perfectly reliable to offset their rarity and incompatibility with mods (9b59163)

`sight_dispersion` and melee properties required few exceptions although in the case of the latter we might want to look at how the class value compare to other objects of similar sizes in a future PR.

Tested with `--check-mods`. The only code change is in 57a53af to ensure `magazines` is cleared when a derivative tries to overwrite that field.

### CODE REVIEW

**Comment by mugling** [comment]
2b98044 drops the `json-format-check` target from `Makefile`. Apart from being very out-of-date it is also reasonable for `copy-from` items to want to re-specify a default value.

A more thorough solution could probably use `jq` but I'm not sure how @narc0tiq would feel adding another dependency to the buildbot?

**Comment by narc0tiq** [comment]
> A more thorough solution could probably use `jq` but I'm not sure how @narc0tiq would feel adding another dependency to the buildbot?

It would be fine by me, as long as v1.3-1.1ubuntu1 is sufficient unto the task.

**Comment by mugling** [comment]
Obviously 2b98044 fails the buildbot as it modifies `Makefile`

File: data/json/items/gun/20x66mm.json [hunk 1]
```diff
@@ -1,26 +1,23 @@
...
     "range": 6,
     "ranged_damage": 4,
     "dispersion": 45,
-    "sight_dispersion": 90,
-    "aim_speed": 4,
     "recoil": 75,
-    "durability": 9,
+    "relative": {
+        "aim_speed": -2,
+	"durability": 1
```
Comment (Coolthulhu) [inline data/json/items/gun/20x66mm.json]: Tab here (and in the shotgun below) make it look like it's out of braces in github display.

**Comment by Coolthulhu** [comment]
Would it be possible to separate the jq/json styling check change into a separate PR?

**Comment by mugling** [comment]
> Would it be possible to separate the jq/json styling check change into a separate PR?

Which part do you want to merge (or both?)

File: data/json/items/gun/20x66mm.json (hunk 1 above)
Comment (mugling) [inline data/json/items/gun/20x66mm.json]: We really need a `json-check` to reject tabs in the JSON to cut down on such errors

**Comment by mugling** [comment]
Gentle bump because the buildbot reports this as failure when it's not.

**Comment by Coolthulhu** [comment]
Does the styling part have to be in this PR? Combining big PRs and PRs that break buildbot verification has a high chance of resulting in problems.

**Comment by mugling** [comment]
> Does the styling part have to be in this PR?

@Coolthulhu No, but it's probably easier to drop the solitary line rejected by `tools/json_format_check.sh` which I've done in 5c3bbf7. If so I'd be grateful for a merge test as I've some more dependent work I'd like to PR tonight if possible.

** REBASED **

**Comment by kevingranade** [comment]
There is a really huge difference between "Uses copy-from to reduce duplication where one gun is a derivative of another. " and "all guns are derivatives of a handful of base types, so effectively no guns have complete definitions".  This approach will very quickly lead to no one being able to determine what the stats of a particular item are by simply reading the json, and I'm very against it.
This is why we don't have inheritance based item definitions in the first place, it's a bad idea because it scatters the definition of a particular item across an arbitrary number of different files.

**Comment by mugling** [comment]
The motivation for item classes comes from @Coolthulhu's suggestion in #15876:

> Also, the skill type hardcode... I'd wait for it till the item variants. Then you could implement it as variants - base item would be an abstract "pistol" or "rifle" which would set the base aim speed.

> "all guns are derivatives of a handful of base types, so effectively no guns have complete definitions"

We've already removed a lot of default values from JSON to improve readability. Providing additional defaults via classes extends this further although I agree there's a balance to be had. For example the already merged `pistol_revolver` class is semantically clearer than explicitly specifying all the required properties individually. Prior to aec2899 many revolvers had inconsistent sets of flags.

> This approach will very quickly lead to no one being able to determine what the stats of a particular item are by simply reading the json

Comparing stats via JSON is already difficult without the use of external tools. Currently items have many fields which can appear in any order with some fields optionally omitted. We also have a large and ever-increasing number of items. As a result many items are poorly balanced.

I can't realistically see anyone actually manually parsing through 100+ guns and checking each property is balanced against every other. Previous attempts have stalled because of the scale of the work required. Even if someone did there would be the ongoing cost of maintaining that balance.  Classes partially obviate this work whilst keeping the game data in a readily editable text fromat.

> This is why we don't have inheritance based item definitions in the first place, it's a bad idea because it scatters the definition of a particular item across an arbitrary number of different files.

One option to limit complexity whilst retaining all of the benefits would be not allowing an item class to inherit from another - effectively forbidding inheritance trees.

**Comment by kevingranade** [comment]
> Prior to aec2899 many revolvers had inconsistent sets of flags.

If a flag is definitional for a group of items, it might be reasonable to
do an inheritance type thing to set a number if flags at once, but if it's
base stats that are being inherited, it's going too far.

> I can't realistically see anyone actually manually parsing through 100+
guns and checking each property is balanced against every other.

Of course not, but the inheritance based definition makes it strictly more
difficult by obfuscation of the stats.  Also, there *is* no strict balance,
the system is *by design* messy and idiosyncratic to capture variation
among various real world items, which is more important to me than trying
to enforce finely tuned balance across all items.

**Comment by Coolthulhu** [comment]
The inheritance design helps a lot with items that might otherwise miss balancing, though. For item sets as big as the set of guns in the current master, manual tweaks generally leave stragglers with old stats. This has happened with dispersion (multiplication by 15) and reloaded ammo rebalance.

Inheritance does what armor value calculations from materials were trying to do, except less badly, with less obfuscation.

Another solution:
* Have both inherited and absolute values in each entry
* Check if they match with buildbot
* Fail the build (but allow manual building) if values don't match
* Possibly allow regenerating the jsons from the inheritance tree with some tool

**Comment by mugling** [comment]
> This has happened with dispersion (multiplication by 15) and reloaded ammo rebalance.

**Comment by mugling** [comment]
> Also, there *is* no strict balance, the system is *by design* messy and idiosyncratic to capture variation among various real world items

Our ability to add variation isn't limited. A STEN gun is `durability: -2` and such syntax is a big win for clarity as it documents intention.

> For item sets as big as the set of guns in the current master, manual tweaks generally leave stragglers with old stats. This has happened with dispersion (multiplication by 15) and reloaded ammo rebalance.

`range` and `durability` are other good examples. The former field is defacto unused with values derived from both ammo and gun and being near-random with some above sight range. `durability` has a similar problem (some guns are for example `durability: 0`) and is a limiting factor to a better implementation of weapon jams/misfires etc A class system could better handle both.

> Inheritance does what armor value calculations from materials were trying to do, except less badly, with less obfuscation.

@kevingranade is correct that there is some obfuscation but the payoff is so much larger. This PR for example makes the improvised weapons usable whereas prior they weren't. Balance for the guns is currently so poor that most aren't used. We can either decrease the number of items, increase the manpower or automate part of the work. The first was discounted some time ago, for the second a cursory scan of `git blame` shows that I'm the only developer actively working on this and the third is this PR.

> Possibly allow regenerating the jsons from the inheritance tree with some tool

I had considered this is but compiling a higher level representation to lower level isn't particularly useful and would be a lot of work versus the amount of use it would likely receive. I'd rather extend `--dump-stats` to be more useful

**Comment by kevingranade** [comment]
> Our ability to add variation isn't limited. A
> STEN gun is durability: -2 and such syntax is
> a big win for clarity as it documents intention.

Good point that it's valid for durability, but for most stats, we aren't
defining gun stats in terms of, "like a default gun, but better/worse X",
we're defining gun stats in absolute terms based on real life stats when
possible, so the relative values are obfscucating rather than clarifying.

Takeaway: let's determine which stats are strictly relative to other
weapons, such as durability, color and symbol, document those as being
mostly derived, and provide the overrides.

>> For item sets as big as the set of guns in the
>> current master, manual tweaks generally
>> leave stragglers with old stats. This has
>> happened with dispersion (multiplication by
>> 15) and reloaded ammo rebalance.

Making reloaded ammo stats be based on factory loads but with penalties is
the original rationale for this feature, and is totally fine. What I'm
objecting to is generalizing this to all guns and ammo.
As for the dispersion rescale, this wouldn't have helped, because all base
and all relative values would have to have been adjusted.

> This PR for example makes the improvised
> weapons usable whereas prior they weren't.
> Balance for the guns is currently so poor that
> most aren't used.

This is a great example of a bad assumption, making all guns desireable to
use is not a valid goal. Some guns are going to only be desireable if you
can't find something better, and that's ok.

**Comment by mugling** [comment]
> This is a great example of a bad assumption, making all guns desireable to use is not a valid goal. Some guns are going to only be desireable if you can't find something better, and that's ok.

I'm actively trying to improve the balance of item spawns but they're still not great - you can always find a good gun before an improvised one and usually on the first day. I'd agree there should be undesirable ones but our current split is 10% good, 90% bad where the bad guns are unremarkable from each other. We need a spectrum from excellent to useless with different trade-offs.

> we're defining gun stats in absolute terms based on real life stats when possible, so the relative values are obfuscating rather than clarifying.

The difficulty is that excluding `weight` and `price` we aren't (and can't) use real world units. A unit of `recoil` is arbitrary and only useful when compared to another item. There's also the problem that all such efforts seem to invariably stall due to the scale of the problem (eg. #15752). As a result we're stuck with unusable values eg. `durability 0`, `range 70`

> As for the dispersion rescale, this wouldn't have helped, because all base and all relative values would have to have been adjusted.

The initial plan is normalization to get at least all values to something sane and then specifying the differences. For example 0432380 makes machine guns usable again.

> Takeaway: let's determine which stats are strictly relative to other weapons, such as durability, color and symbol, document those as being mostly derived, and provide the overrides.

A specification is always good.

I'm acutely aware that we're trying to release `0.D` and I'd like the work I've done on the guns over the preceding year to be in a more polished state.

Can I propose that we define our long-term aim for field specification but go with relative values for now. It's possible for me to balance the guns prior to `0.D` that way by specifying defaults and deviations whereas the other approach is a 4000 line PR and has little hope of being completed for then. I'd be disappointed having done all the prior work not to get the implementation balanced and working for the release.

This PR was reworked to ensure everything remains in JSON, that absolute values can still be specified and that load-order is irrelevant. Accordingly we can gradually move to absolute values over time without penalty. As the original JSON values are badly broken there isn't any loss to this approach and if we have a specification we can work through it in stages whilst still having usable guns for `0.D`

**Comment by kevingranade** [comment]
>> This is a great example of a bad assumption, making all guns desireable to use is not a valid goal. Some guns are going to only be desireable if you can't find something better, and that's ok.

> I'm actively trying to improve the balance of item spawns but they're still not great - you can always find a good gun before an improvised one and usually on the first day. I'd agree there should be undesirable ones but our current split is 10% good, 90% bad where the bad guns are unremarkable from each other. We need a spectrum from excellent to useless with different trade-offs.

This is still making a strong and unwarranted assumption about the role of improvised weapons, guns in particular.  It IS the case that it is reasonable to find real factory built guns very quickly in this scenario.  It is also the case that improvised guns will be quite worthless compared to factory-made guns. Furthermore there are a number of factory-made guns that will be strictly inferior to the majority of other guns.  Trying to make all guns competitive with each other is not something we should be trying to do.  Each gun should exemplify it's real-world counterpart as well as it can, and that's all there is to it.

>> we're defining gun stats in absolute terms based on real life stats when possible, so the relative values are obfuscating rather than clarifying.

> The difficulty is that excluding weight and price we aren't (and can't) use real world units. A unit of recoil is arbitrary and only useful when compared to another item. There's also the problem that all such efforts seem to invariably stall due to the scale of the problem (eg. #15752). As a result we're stuck with unusable values eg. durability 0, range 70

We absolutely can, a number of stats (damage, recoil, noise) are derivable from muzzle velocity, bullet mass, and weapon mass.  There are figures for range and report volume for quite a lot of guns, and those are just off the top of my head.

>> As for the dispersion rescale, this wouldn't have helped, because all base and all relative values would have to have been adjusted.

> The initial plan is normalization to get at least all values to something sane and then specifying the differences. For example 0432380 makes machine guns usable again.

My point was that the "class" system would not have prevented the problem of values getting out of sync.

> Can I propose that we define our long-term aim for field specification but go with relative values for now. 

As I've pointed out a number of times in this exchange, for game-specific and arbitrary traits, yes, but for stats based on real-world gun performance, no.  If you feel a particular stat is not possible to derive from the real gun, make an argument for that, but the default should be to use real data whenever possible.  For example, I'd absolutely love to have dispersion based on some kind of real gun data, but there seems to be pretty much zero data on gun accuracy available, and there are really no contributing stats that we can base it on.  Damage on the other hand is very much an absolute value based on properties of the gun and ammunition, and shouldn't be defined as relative to other guns or ammo.

> It's possible for me to balance the guns prior to 0.D that way by specifying defaults and deviations whereas the other approach is a 4000 line PR and has little hope of being completed for then. I'd be disappointed having done all the prior work not to get the implementation balanced and working for the release.

I'm not sure this is a feasible goal in the first place, it sounds like you want to overhaul the balance of basically every firearm in the game, that is not something to try and fit in right before a release.  If anything it's precisely the kind of thing I'd prefer to delay until right after a release to give it maximum time to settle back down before the next.
Also, as demonstrated by the improvised guns sub-discussion, we are not agreeing in the slightest about what that balance should look like, so that makes it even more risky.

**Comment by BorkBorkGoesTheCode** [comment]
> For example, I'd absolutely love to have dispersion based on some kind of real gun data, but there seems to be pretty much zero data on gun accuracy available, and there are really no contributing stats that we can base it on. 

If you're looking for gun dispersion, have you been searching for M.O.A. data? Also, have you looked up shooting competition data? I'm pretty sure you can find info on the specific guns used and the shot groupings achieved.

**Comment by kevingranade** [comment]
> If you're looking for gun dispersion, have you been searching for M.O.A. data?

Yes, unsuccessfully. The only things I came across the last time I looked were people arguing on forums about whether particular guns are sub-MoA or not.

> Also, have you looked up shooting competition data? I'm pretty sure you can find info on the specific > guns used and the shot groupings achieved.

I had not thought of this approach, it sounds like a good one.

**Comment by kevingranade** [comment]
Reiterating: let's determine which stats are strictly relative to other weapons, such as durability, color and symbol, document those as being mostly derived, and provide the overrides.

Until then, closing.

**Pull request closed without merging**